File: src/bot/listeners/messageListener.ts
import { Client } from 'tdl';
import { messageDocument, updateNewMessage } from 'tdlib-types';
import { isChat } from '../filters';
import { debug } from '../helper';
import { documentHandler } from './contents/documentHandler';
import { photoHandler } from './contents/photoHandler';
import { videoMessageHandler } from './contents/videoMessageHandler';

const CONTENT_HANDLERS: Record<
  string,
  (update: updateNewMessage, client: Client) => void
> = {
  messageVideoNote: (update, client) => videoMessageHandler(update, client),
  messagePhoto: (update, client) => photoHandler(update, client),
  messageDocument: (update, client) =>
    documentHandler(update, client, update.message.content as messageDocument),
};

const defaultHandler = (update: updateNewMessage) => {
  debug(
    'Received content',
    update.message.content._,
    JSON.stringify(update.message.content)
  );
};

export const messageListener = (
  update: updateNewMessage,
  client: Client
): void => {
  // only accept specific chat
  if (!isChat(update)) {
    debug('Received update from other chat', JSON.stringify(update));
    return;
  }

  console.log('Received new message');

  const content = update.message.content;
  // based on content
  const handler = CONTENT_HANDLERS[content._] ?? defaultHandler;

  handler(update, client);
};

File: src/main/java/com/pirasalbe/services/telegram/listeners/MessageListener.java
package com.pirasalbe.services.telegram.listeners;

import java.util.function.BiConsumer;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import com.pirasalbe.configurations.TelegramConfiguration;
import com.pirasalbe.helpers.FilterHelper;
import com.pirasalbe.services.telegram.listeners.contents.DocumentHandler;
import com.pirasalbe.services.telegram.listeners.contents.PhotoHandler;
import com.pirasalbe.services.telegram.listeners.contents.VideoMessageHandler;

import it.tdlight.client.SimpleTelegramClient;
import it.tdlight.jni.TdApi.MessageContent;
import it.tdlight.jni.TdApi.MessageDocument;
import it.tdlight.jni.TdApi.MessagePhoto;
import it.tdlight.jni.TdApi.MessageVideoNote;
import it.tdlight.jni.TdApi.UpdateNewMessage;

@Component
public class MessageListener implements BiConsumer<SimpleTelegramClient, UpdateNewMessage> {

	private static final Logger LOGGER = LoggerFactory.getLogger(MessageListener.class);

	@Autowired
	private TelegramConfiguration configuration;

	@Autowired
	private PhotoHandler photoHandler;
	@Autowired
	private DocumentHandler documentHandler;
	@Autowired
	private VideoMessageHandler videoMessageHandler;

	@Override
	public void accept(SimpleTelegramClient client, UpdateNewMessage update) {
		if (!FilterHelper.isChat(configuration.getChatId(), update)) {
			LOGGER.debug("Received update from other sender {}", update);
			return;
		}

		LOGGER.info("Received new message");

		MessageContent content = update.message.content;
		if (content instanceof MessagePhoto) {
			photoHandler.accept(client, update);
		} else if (content instanceof MessageDocument) {
			documentHandler.accept(client, update, (MessageDocument) content);
		} else if (content instanceof MessageVideoNote) {
			videoMessageHandler.accept(client, update);
		} else {
			LOGGER.debug("Received content {}", update.message.content);
		}
	}

}
